arbor/cable_cell.cpp

```diff
@@ -54,41 +148,210 @@ cable_segment* cable_cell::add_cable(index_type parent, segment_ptr&& cable) {
+
+const std::vector<index_type>& cable_cell::parents() const {
+    return impl_->parents;
+}
+
+value_type cable_cell::segment_length_constant(value_type frequency, index_type segidx,
+    const cable_cell_parameter_set& global_defaults) const
+{
+    return 0.5/segment_mean_attenuation(frequency, segidx, global_defaults);
 }
 
 bool cable_cell::has_soma() const {
     return !segment(0)->is_placeholder();
 }
 
-void cable_cell::paint(const std::string& target, const std::string& description) {
-    auto it = regions_.find(target);
+const std::vector<cable_cell::gap_junction_instance>& cable_cell::gap_junction_sites() const {
+    return impl_->gap_junction_sites;
+}
+
+const std::vector<cable_cell::synapse_instance>& cable_cell::synapses() const {
+    return impl_->synapses;
+}
+
+const std::vector<cable_cell::detector_instance>& cable_cell::detectors() const {
+    return impl_->spike_detectors;
+}
+
+const std::vector<cable_cell::stimulus_instance>& cable_cell::stimuli() const {
+    return impl_->stimuli;
+}
+
+void cable_cell::set_regions(cable_cell::region_map r) {
+    impl_->regions = std::move(r);
+}
+
+void cable_cell::set_locsets(cable_cell::locset_map l) {
+    impl_->locations = std::move(l);
+}
+
+//
+// Painters.
+//
+// Implementation of user API for painting density channel and electrical properties on cells.
+//
+
+void cable_cell::paint(const std::string& target, mechanism_desc desc) {
+    auto it = impl_->regions.find(target);
 
     // Nothing to do if there are no regions that match.
-    if (it==regions_.end()) return;
+    if (it==impl_->regions.end()) return;
 
-    for (auto i: it->second) {
-        segment(i)->add_mechanism(description);
+    for (auto c: it->second) {
+        if (c.prox_pos!=0 || c.dist_pos!=1) {
+            throw cable_cell_error(util::pprintf(
+                "cable_cell does not support regions with partial branches: \"{}\": {}",
+                target, c));
+        }
+        segment(c.branch)->add_mechanism(std::move(desc));
     }
 }
 
+//
+// Placers.
+//
+// Implementation of user API for placing discrete items on cell morphology,
+// such as synapses, spike detectors and stimuli.
+//
+
+//
+// Synapses
+//
+
+locrange cable_cell::place(const std::string& target, const mechanism_desc& desc) {
+    const auto first = impl_->synapses.size();
+
+    const auto it = impl_->locations.find(target);
+    if (it==impl_->locations.end()) return locrange(first, first);
+
+    return impl_->place(it->second, desc, impl_->synapses);
+}
+
+/*
+locrange cable_cell::place(const locset& ls, const mechanism_desc& desc) {
+    const auto locs = thingify(ls, impl_->morph);
+    return impl_->place(locs, desc, impl_->synapses);
+}
+*/
+
+locrange cable_cell::place(const mlocation& loc, const mechanism_desc& desc) {
+    if (!test_invariants(loc) || loc.branch>=num_segments()) {
+        throw cable_cell_error(util::pprintf(
+            "Attempt to add synapse at invalid location: \"{}\"", loc));
+    }
```

Comment:
Until there's always an embedded morphology to check against, could these location validity tests be provided by a single routine in `impl_`, again to avoid some boilerplate? e.g. via a new member function `assert_valid_location`.

Reply:
It doesn't save much, because we have a custom message for each `place` call.
I have added a `bool valid_location(mlocation)` function to perform the test.

Reply:
I don't think we need a different message for each type of place error though.

Reply:
Another note: all of this code will change later with generic expressions for place calls, so no need to sweat details about error messages now.